DigitalLearningSolutions.Data.Migrations/202204071530_PopulateUsersTableFromAccountsTables.cs

```diff
@@ -0,0 +1,415 @@
+                        SELECT PrimaryEmail FROM Users)"
+            );
+            connection.Execute(
+                @"UPDATE DelegateAccounts
+                    SET
+                        UserID = (SELECT ID FROM Users WHERE Email = PrimaryEmail),
+                        Email = NULL,
+                        CentreSpecificDetailsLastChecked = GETUTCDATE()"
+            );
+
+            // Get the rest of the delegate accounts we've not resolved yet
+            var delegateAccounts =
+                connection.Query<DelegateAccount>("SELECT * FROM DelegateAccounts WHERE UserId IS NULL");
+
+            var delegateAccountsGroupedByEmail = delegateAccounts.GroupBy(da => da.Email);
+
+            foreach (var emailGroup in delegateAccountsGroupedByEmail)
+            {
+                if (!string.IsNullOrWhiteSpace(emailGroup.Key))
+                {
+                    var delegatesToAttemptToLink = new List<DelegateAccount>();
+                    var delegateAccountsGroupedByCentre = emailGroup.GroupBy(da => da.CentreId);
+
+                    foreach (var centreGroup in delegateAccountsGroupedByCentre)
+                    {
+                        delegatesToAttemptToLink.Add(centreGroup.First());
+
+                        // All duplicate emails at a centre need new User accounts.
+                        // If there are more than one delegate accounts with the same email at a centre
+                        // we only want to keep the email on the first. The rest get reset to having a User with a guid email
+                        var othersAtCentre = centreGroup.Skip(1);
+
+                        foreach (var delegateAccount in othersAtCentre)
+                        {
+                            // Insert a new User with Guid email
+                            InsertNewUserForDelegateAccount(connection, delegateAccount, true);
+                        }
+                    }
+
+                    var allJobGroupsMatch =
+                        delegatesToAttemptToLink.Select(da => da.JobGroupId_deprecated).Distinct().Count() < 2;
+                    foreach (var delegateAccount in delegatesToAttemptToLink)
+                    {
+                        // Check for existing user with email
+                        var existingUserWithEmail = connection.QuerySingleOrDefault<User>(
+                            "SELECT * FROM Users WHERE PrimaryEmail = @email",
+                            new { email = emailGroup.Key }
+                        );
+
+                        if (existingUserWithEmail != null)
+                        {
+                            // If we find a user, we update any default data on that user
+                            UpdateExistingUserWithDelegateDetails(
+                                connection,
+                                existingUserWithEmail,
+                                delegateAccount,
+                                allJobGroupsMatch

```

Comment:
OLD COMMENT: Shouldn't we be checking not just the jobGroups match, but that all first name, last name, profile image, PRN details, are the same across all DelegateAccounts. Otherwise we are just selecting a seemingly random set of details from one of the accounts to update with? Looks like we are going to updating the single User record with details from each of the DelegateAccounts.

NEW COMMENT AFTER READING THROUGH MORE OF THE CODE: I think this method name could be improved. As you can see from my old comment, I was getting very confused as it looked like we kept overwriting details with other ones. Mostly because the method name does not highlight that differing details won't be overwritten, and exacerbated by the fact we check the jobGroups match and none of the others. 

Why are we checking the jobGroups before calling the method? I feel like they can be done inside it along with the others.

Reply:
Because we are supposed to check all the jobgroups according to the ticket. @SteveJacksonSoft Should we be changing something here?

Reply:
I only mention the jobGroups bit because we have this whopping great method to check all the various details match, but the jobGroups are checked outside of it. From my naïve look, it seemed like that check could have been moved inside the method, as it didn't seem to be doing anything different from the others.

Reply:
https://github.com/TechnologyEnhancedLearning/DLSV2/pull/1065#discussion_r847457945 another relevant comment chain

Reply:
The logic here is right. We should set the user detailsLastChecked to null if there are any conflicts in the fields that we're merging. Most of those fields are on both delegateAccounts and adminAccounts, but jobGroup is only on delegateAccounts. Because of the way we're implementing the data migration, it's easiest to check the former iteratively and the latter all at once.
An improvement would be to pull `allJobGroupsMatch` out of `DoesDelegateAccountMatchExistingUser()` and combine the two with a && to calculate `@detailsMatch` in the User update query. But I don't _think_ it's worth doing anything more than that.

Reply:
Could maybe explain why `allJobGroupsMatch` is different in a comment I guess.